lib/driver.js

```diff
@@ -1090,7 +1215,100 @@ export class XCUITestDriver extends BaseDriver {
     if (cmd === 'getStatus') {
       return await this.getStatus();
     }
-    return await super.executeCommand(cmd, ...args);
+
+    if (cmd === 'createSession') {
+      return await super.executeCommand(cmd, ...args);
+    }
+
+    const wdaRestartEnabled = Boolean(this.opts.enableWdaRestart);
+
+    if (!wdaRestartEnabled) {
+      return await super.executeCommand(cmd, ...args);
+    }
+
+    const maxAttempts = this.opts.wdaStartupRetries || 2;
+    let attempts = 0;
+
+    while (true) {
+      try {
+        return await super.executeCommand(cmd, ...args);
+      } catch (err) {
+        this.log.debug(`Error executing command '${cmd}': ${err.message}`);
+        attempts++;
+
+        const isWdaCrash = this.isWdaCrashOrSessionInvalid(err);
+
+        if (!isWdaCrash || attempts > maxAttempts) {
+          throw err;
+        }
+
+        this._isRecovering = true;
+        this.log.warn(`WDA issue detected. Recovery attempt ${attempts}/${maxAttempts}...`);
+
+        try {
+          await this.handleWdaRecovery();
+          this.log.info('Recovery successful, retrying original command');
+        } catch (recoveryErr) {
+          this.log.error(`Recovery attempt failed: ${recoveryErr.message}`);
+          if (attempts >= maxAttempts) {
+            throw new Error(
+              `Failed to recover WDA after ${maxAttempts} attempts: ${recoveryErr.message}`,
+            );
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Checks if an error indicates a WDA crash or invalid session
+   * @private
+   * @param {Error} err - The error to check
+   * @returns {boolean} - True if the error indicates WDA crash or invalid session
+   */
+  isWdaCrashOrSessionInvalid(err) {
+    if (this.isWdaConnectionError(err)) {
+      return true;
+    }
+
+    return Boolean(
+      err.message &&
+        (err.message.includes('invalid session id') ||
+          err.message.includes('Session does not exist')),
```

Comment:
These messages are managed in https://github.com/appium/WebDriverAgent (WDA code).

To keep error message management in a closed place, probably this kind of error check should be in https://github.com/appium/WebDriverAgent/tree/master/lib. So, as xcuitest driver, https://github.com/appium/appium-xcuitest-driver/pull/2531/files#diff-ed3c013edbc896bd87ef62539626fcdc8930a31f3cee7ecb37438e172af2c8e0R1291-R1303 these lines can be in the WDA JS side and XCTest driver could call it as `this.wda.hasCrashedOrSessionInvalidError()`. If true, xcuitest driver can do the recovery steps. This is similar to `this.wda.isRunning()`